stable_baselines3/common/buffers.py

```diff
@@ -188,7 +210,8 @@ def _get_samples(self,
                      ) -> ReplayBufferSamples:
         data = (self._normalize_obs(self.observations[batch_inds, 0, :], env),
                 self.actions[batch_inds, 0, :],
```

Comment:
```suggestion
                     ) -> ReplayBufferSamples:
        batch_inds = (batch_inds + pos) % self.buffer_size
        data = (self._normalize_obs(self.observations[batch_inds, 0, :], env),
                self.actions[batch_inds, 0, :],
```

This corrects the sample alignment issue we discussed when using a single array of observations.

Reply:
Alternatively, override the sample function with this:

```python
    def sample(self,
               batch_size: int,
               env: Optional[VecNormalize] = None
               ):
        """
        :param batch_size: (int) Number of element to sample
        :param env: (Optional[VecNormalize]) associated gym VecEnv
            to normalize the observations/rewards when sampling
        :return: (Union[RolloutBufferSamples, ReplayBufferSamples])
        """
        upper_bound = self.buffer_size if self.full else self.pos
        batch_inds = (np.random.randint(1, upper_bound, size=batch_size)+self.pos) % self.buffer_size
        return self._get_samples(batch_inds, env=env)
```

Reply:
ok, I will see, need to think at it with a clear head (so tomorrow ^^#)

Reply:
Of course. I will add some tests that verify this specific issue.

Reply:
the first solution does not seem to work but the second does ;)

Reply:
Woups, `self.pos` haha. I think the latter is cleaner ^^'

Reply:
The only caveat here is that the implementation expects that the trajectories are continuous which, I suppose, is fine for now.

Reply:
I am rather confused by this function. If we only have samples till `self.pos` index, and take random integers from [`self.pos + 1`, `self.pos + 1 + self.pos`], won't some of these be above `self.pos` and thus lead to invalid observations?

Reply:
and it should be 
```python
batch_inds = (np.random.randint(1, upper_bound, size=batch_size) + self.pos) % upper_bound
```
otherwise you are going to sample a lot of zeros...

Reply:
actually, still not correct

Reply:
I think this is still same as just doing `batch_inds = np.random.randint(1, upper_bound, size=batch_size)`. I am not even sure what we are trying to fix here any more.

Reply:
@Miffyli , yes that only works when the buffer is full.

This works for both cases:

```python
    def sample(self,
               batch_size: int,
               env: Optional[VecNormalize] = None
               ):
        """
        :param batch_size: (int) Number of element to sample
        :param env: (Optional[VecNormalize]) associated gym VecEnv
            to normalize the observations/rewards when sampling
        :return: (Union[RolloutBufferSamples, ReplayBufferSamples])
        """
        
        upper_bound = self.buffer_size if self.full else self.pos
        
        if self.full:
             batch_inds = (np.random.randint(1, self.buffer_size, size=batch_size)+self.pos) % self.buffer_size
        else:
              batch_inds = (np.random.randint(1, self.pos, size=batch_size)+self.pos) % self.buffer_size
        return self._get_samples(batch_inds, env=env)
```

The first part samples from {0.. buffer_size-1} - {pos}, the second is for the case where the buffer isn't full, so sampling {0..pos-1} is fine.

Reply:
The whole point is to avoid sampling `self.pos` because once the buffer is full, then sampling from the index `self.pos` is wrong, because that will use two unrelated observations.

Reply:
This notebook illustrates the problems with current implementation. The source is copied and pasted from the repo:

https://colab.research.google.com/drive/1J_9TVXdPTPemFv8sPxzYwgqb6tGfyhCx?usp=sharing

Reply:
First, with an array of size `buff_size`, we have access to buff_size - 1 observations.
Second, sampling `self.pos` should never be possible, but
`batch_inds = np.random.randint(0, upper_bound, size=batch_size)` from the base class doesn't account for `self.pos` when the buffer is full, therefore, at some point, it will sample `self.pos` if the buffer is full.

Reply:
>This notebook illustrates the problems with current implementation. 

I don't have access to it... but implemented your last solution locally (tests are running)

Reply:
Ah sorry. This should do it:

https://colab.research.google.com/drive/1J_9TVXdPTPemFv8sPxzYwgqb6tGfyhCx?usp=sharing